### PR TITLE
Assignment 3

### PR DESCRIPTION
**Pull request by gknutson1**

### CODE REVIEW

**Comment by BenagainB** [comment]
Assignment
For this assignment do the following:

Create two interfaces. One to represent the JokeService and one interface to represent outputting the joke to the screen (Console.WriteLine). ✔
The JokeService will need to have the interface applied to it. ✔
Create an implementation of the output interface that writes out the joke to the console. ✔
Implement the Jester class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
The Jester class TellJoke() method should retrieve a joke from the JokeService. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔
Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ✔
Fundamentals
Be sure you enable:
Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ❌
For this assignment, always use Assert.AreEqual<T>() (the generic version) ❌
Ensure there are no errors or warnings (including code analysis warnings) ❌
All of the above should be unit tested ❌
Choose simplicity over complexity ✔
Screen Shot 2023-01-31 at 10 52 19 PM
Screen Shot 2023-01-31 at 10 54 01 PM
-images.githubusercontent.com/91215159/215974028-32f9446d-6484-405c-9cd1-5fb0449d79f8.png">
I'm seeing 5 warnings from nullability issues you should resolve prior to final turn in.

Take another look at your tests using Assert.AreEqual and compare that to the required generic usage before the final turn in.

Overall this looks very good and, if the string parser is revised to handle quotation marks, the user would be satisfied with the output.

**Comment by JessicaH46** [comment]
I will review this code

File: CanHazFunny/CanHazFunny.Tests/CanHazFunny.Tests.csproj [hunk 1]
```diff
@@ -1,9 +1,10 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>net6.0</TargetFramework>
-
+    <TargetFramework>net7.0</TargetFramework>
     <IsPackable>false</IsPackable>
+    <Nullable>enable</Nullable>
+    <LangVersion>11</LangVersion>
```
Comment (JessicaH46) [inline CanHazFunny/CanHazFunny.Tests/CanHazFunny.Tests.csproj:7]: <EnableNETAnalyzers>true</EnableNETAnalyzers>

Comment (JessicaH46) [reply CanHazFunny/CanHazFunny.Tests/CanHazFunny.Tests.csproj:7]: Need this one too, I'm pretty sure
`<WarningsAsErrors>true</WarningsAsErrors>`

**Review by JessicaH46 — COMMENTED** [review]
Create two interfaces. One to represent the JokeService and one interface to represent outputting the joke to the screen (Console.WriteLine). ✔
The JokeService will need to have the interface applied to it. ✔
Create an implementation of the output interface that writes out the joke to the console. ✔
Implement the Jester class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
The Jester class TellJoke() method should retrieve a joke from the JokeService. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔
Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ✔
Fundamentals
Be sure you enable:
Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ❌
For this assignment, always use Assert.AreEqual<T>() (the generic version) ❌
Ensure there are no errors or warnings (including code analysis warnings) ❌
All of the above should be unit tested ❌
Choose simplicity over complexity ✔


Going through your code I think you forgot to do the Net analyzers so I left a comment there, I also realized that I messed up the generic Assert.AreEqual <T>, as someone mentioned about your code above.

**Review by DaisyPalencia — COMMENTED** [review]
Create two interfaces. One to represent the JokeService and one interface to represent outputting the joke to the screen (Console.WriteLine). ✔
The JokeService will need to have the interface applied to it. ✔
Create an implementation of the output interface that writes out the joke to the console. ✔
Implement the Jester class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
The Jester class TellJoke() method should retrieve a joke from the JokeService. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔
Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ✔
Fundamentals
Be sure you enable:
Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ❌
For this assignment, always use Assert.AreEqual() (the generic version) ❌
Ensure there are no errors or warnings (including code analysis warnings) ❌
All of the above should be unit tested ❌
Choose simplicity over complexity ✔

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+public class Jester
+{
+    private readonly IJokeService _jokeService;
+    private readonly IOutput _output;
+
+    public Jester(IOutput? output, IJokeService? jokeService)
+    {
+        _output = output ?? throw new ArgumentNullException(nameof(output));
+        _jokeService = jokeService ?? throw new ArgumentNullException(nameof(jokeService));
+        _jokeService = jokeService ?? throw new ArgumentNullException(nameof(jokeService));
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/Jester.cs:14]:
```suggestion
```
Just need to do this once. I assume this is a typo

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    }
+
+    public void TellJoke()
+    {
+        string joke;
+        while ((joke = _jokeService.GetJoke()).ToLower().Contains("chuck norris")) {}
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/Jester.cs:20]:
```suggestion
        string joke = _jokeService.GetJoke();
        while (jokeToLower().Contains("chuck norris")) {
          joke = _jokeService.GetJoke();
        }
```

While what you're doing works, I'd argue this is easier to read and understand.

**Pull request closed without merging**